File: bintree_py/binarytree/traveler.py
'''

BINARY TREE TRAVELER

Description:    Traverses through binary trees.

Author:         Thanh Trung Nguyen
                thanh.it1995 (at) gmail.com

License:        3-Clause BSD License

'''


#
#
class Traveler:
    def traverse(self, root, order: str) -> list:
        pass


#
#
class TravelerRecur(Traveler):
    #
    #
    def traverse(self, root, order: str) -> list:
        '''
        Traverses a tree entirely.
        Args:
            order: Type of traversal order.
                - 'pre': pre-order
                - 'in': in-order
                - 'post': post-order
        Returns:
            A list of keys.
        '''

        if order not in ('pre', 'in', 'post'):
            raise ValueError('Invalid argument: order')

        self._res_path = []

        if order == 'pre':
            self._traverse_pre(root)
        elif order == 'in':
            self._traverse_in(root)
        elif order == 'post':
            self._traverse_post(root)

        res = self._res_path
        del self._res_path

        return res

    #
    #
    def _traverse_pre(self, node):
        if node is None:
            return

        self._res_path.append(node.key)
        self._traverse_pre(node.left)
        self._traverse_pre(node.right)

    #
    #
    def _traverse_in(self, node):
        if node is None:
            return

        self._traverse_in(node.left)
        self._res_path.append(node.key)
        self._traverse_in(node.right)

    #
    #
    def _traverse_post(self, node):
        if node is None:
            return

        self._traverse_post(node.left)
        self._traverse_post(node.right)
        self._res_path.append(node.key)

File: bintree_java/src/binarytree/Traveler.java
package binarytree;


import java.util.List;



public abstract class Traveler< TKey extends Number & Comparable<? super TKey>,
                                TNode extends BinNode<TKey,TNode>
                              >
{
    public abstract void traverse(TNode root, OrderTraverse order, List<?> result);
}



//////////////////////////////////////////////////////////////



class TravelerRecur< TKey extends Number & Comparable<? super TKey>,
                     TNode extends BinNode<TKey,TNode>
                   >
      extends Traveler<TKey, TNode>
{


    protected List<TKey> resPath;



    @SuppressWarnings("unchecked")
    @Override
    public void traverse(TNode root, OrderTraverse order, List<?> result) {
        resPath = (List<TKey>)result;
        resPath.clear();

        switch (order) {
        case PRE:
            traversePre(root);
            break;

        case IN:
            traverseIn(root);
            break;

        case POST:
            traversePost(root);
            break;

        default:
            break;
        }
    }



    protected void traversePre(TNode node) {
        if (null == node)
            return;

        resPath.add(node.key);
        traversePre(node.left);
        traversePre(node.right);
    }



    protected void traverseIn(TNode node) {
        if (null == node)
            return;

        traverseIn(node.left);
        resPath.add(node.key);
        traverseIn(node.right);
    }



    protected void traversePost(TNode node) {
        if (null == node)
            return;

        traversePost(node.left);
        traversePost(node.right);
        resPath.add(node.key);
    }
}
